feat: add Casbin-based access control

kushagra189 opened the pull request:
## Description

- Access control support for `foca` using Casbin
- Casbin mongo DB adapter implementation for saving permissions
- API specs definition and controllers for handling permission endpoints
- Decorator for enabling access control over specific endpoints
- Access control-based petstore example
- Default casbin policy definitions

Fixes #107 

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## Checklist:

- [x] My code follows the [style guidelines](https://github.com/elixir-cloud-aai/elixir-cloud-aai/blob/dev/resources/contributing_guidelines.md#language-specific-guidelines) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not reduced the existing code coverage

Commit:
added permission specs, config

Commit:
added permission specs, config

Commit:
fix base image config under petstore app

kushagra189 committed:
functional flow  permission management swagger ui

kushagra189 committed:
initialize controller definitions

kushagra189 committed:
added pymongo adapter

kushagra189 committed:
fix mongo connection specs

kushagra189 committed:
add delete endpoint and logs

kushagra189 committed:
completed endpoints

uniqueg commented on examples/petstore/docker-compose.yaml:
```diff
@@ -1,8 +1,15 @@
+    build:
+      context: ../../
+      dockerfile: docker/Dockerfile_py3.9
+    image: foca-petstore-root:latest
```
To make it more clear what is happening here, perhaps add a comment to say that this is a workaround to get that image built based on the current state of the repo and that after the image is built the container is started but immediately stops. You could also add an explicit `restart: no` (it's the default, but it's easier to understand/follow if stated explicitly)

kushagra189 replied:
marking as done.

uniqueg commented on examples/petstore/config.yaml:
```diff
@@ -42,3 +42,7 @@ db:
                 pets:
                     indexes: null
+
+access:
```
I think `access` and `enable` is misleading, because access is actually the least restricted if `enable` is `false`, which is counter-intuitive. I would probably prefer to have `access` renamed to `access_control`.

kushagra189 replied:
marking as done.

uniqueg commented on examples/petstore/petstore_policies.conf:
```diff
@@ -0,0 +1,15 @@
+# Request definition
```
Given that this is an example it would be good to add comments that explain in more detail what is happening here in each line, with a link to more information on Casbin/PyCasbin

kushagra189 replied:
done.

uniqueg replied:
I don't see where this is done? I was thinking of more expressive/explicit/detailed descriptions, explaining what each of these statements mean

uniqueg replied:
Never mind, I saw this now

uniqueg commented on foca/database/register_mongodb.py:
```diff
@@ -30,7 +30,6 @@ def register_mongodb(
     # Iterate over databases
     if conf.dbs is not None:
         for db_name, db_conf in conf.dbs.items():
-
```
I think it's better not to change more files than necessary in each PR, especially if they are rather big ones, and so given that the line removal here is the only change and has no functional consequences, I would put that line back in again

kushagra189 replied:
marking as done.

uniqueg commented on foca/foca.py:
```diff
@@ -12,6 +12,8 @@
 from foca.factories.connexion_app import create_connexion_app
 from foca.factories.celery_app import create_celery_app
 from foca.security.cors import enable_cors
+from foca.permission_management.config_utils import _create_permission_config, _register_permission_specs, _register_casbin_enforcer
```
It's best practice to put multiple imports in parentheses, and if the lines becomes long, to put individual imports on different lines, e.g.,:

```
from foca.permission_management.config_utils import (
    _create_permission_config.
    _register_permission_specs,
    _register_casbin_enforcer,
)

etc. Similarly for the next line.

Also, there is a Python convention that methods/function whose names start with an underscore are to be considered private (they aren't _really_ private, because there is no such thing in Python). So if you have functions that you are going to import elsewhere in your code (except within your test cases), you probably shouldn't have their names start with a `_`.

uniqueg replied:
I'm also thinking that in order to stay consistent with the corresponding section in the app config, it might be better to rename the module `access_control` rather than `permission_management`. It's also 50% shorter...

kushagra189 replied:
marking as done.

uniqueg commented on foca/models/config.py:
```diff
@@ -1094,6 +1094,39 @@ class LogConfig(FOCABaseConfig):
     root: Optional[LogRootConfig] = LogRootConfig()
 
 
+class AccessConfig(FOCABaseConfig):
```
As mentioned above, better `AccessControlConfig`

kushagra189 replied:
done.

uniqueg commented on foca/models/config.py:
```diff
@@ -1094,6 +1094,39 @@ class LogConfig(FOCABaseConfig):
+    owner_headers: Optional[set] = {'X-User', 'X-Group'}
+    user_headers: Optional[set] = {'X-User'}
+
+    #TODO: add check for a valid conf file
```
Please add issues for these and remove the `TODO`s from the code

kushagra189 replied:
resolved.

uniqueg commented on foca/models/config.py:
```diff
@@ -1094,6 +1094,39 @@ class LogConfig(FOCABaseConfig):
+    #TODO: add check for a valid conf file
+    #TODO: should we support multiple conf files?
+
+    # resolve relative path
```
Is there something missing? If yes, please implement, if not, please remove comment/stub

kushagra189 replied:
resolved

uniqueg commented on foca/models/config.py:
```diff
@@ -1158,6 +1191,7 @@ class 'werkzeug.exceptions.BadGateway'>: {'title': 'Bad Gateway', 'status': 50\
     db: Optional[MongoConfig] = None
     jobs: Optional[JobsConfig] = None
     log: LogConfig = LogConfig()
+    access: Optional[AccessConfig] = None
```
Please update docstring (`Args`, `Attributes` and `Example`). Please update examples also in dependent models, if any (probably there aren't any though).

kushagra189 replied:
done.

uniqueg commented on foca/models/config.py:
```diff
@@ -1094,6 +1094,39 @@ class LogConfig(FOCABaseConfig):
+
+    Args:
+        enable: Flag to enable/disable access control.
+        path_out: Policy file based on `pycasbin` specification.
```
Might be good to add a link here and in the corresponding attribute, something like, `Cf. https://github.com/casbin/pycasbin`

kushagra189 replied:
done

uniqueg commented on foca/models/config.py:
```diff
@@ -1094,6 +1094,39 @@ class LogConfig(FOCABaseConfig):
+    Args:
+        enable: Flag to enable/disable access control.
+        path_out: Policy file based on `pycasbin` specification.
+
```
Please also add/describe `owner_headers` and `user_headers` if you use them in your code. If you are _not_ making use of them in your code, please remove them from the model.

kushagra189 replied:
done

uniqueg commented on foca/models/config.py:
```diff
@@ -1094,6 +1094,39 @@ class LogConfig(FOCABaseConfig):
+    """
+    enable: bool = False
+    policy_path: Optional[str] = None
+    owner_headers: Optional[set] = {'X-User', 'X-Group'}
```
Wouldn't it be better to set the defaults to `None` if you allow these to be optional?

kushagra189 replied:
done

uniqueg commented on foca/foca.py:
```diff
@@ -59,6 +64,12 @@ def foca(config: Optional[str] = None) -> App:
     else:
         logger.info(f"No OpenAPI specifications provided.")
+        
+    if conf.access and conf.access.enable:
```
Would be better to be more explicit about `if conf.access`: the property `.access` should always be there in a config instance, but it _may_ be `None` - which is what you are checking here. It's not good practice to use `None` in boolean context: https://stackoverflow.com/questions/7816363/if-a-vs-if-a-is-not-none

So better/safer to say: `if conf.access is not None and conf.access.enable:` here...

kushagra189 replied:
done

uniqueg commented on foca/foca.py:
```diff
@@ -36,6 +38,9 @@ def foca(config: Optional[str] = None) -> App:
         logger.info(f"Default app configuration used.")
 
+    # Add permission specs
+    conf = _create_permission_config(conf)
```
Do you need to do this if `conf.access.enable` is `False`? Also, please add logging.

kushagra189 replied:
done.

uniqueg commented on foca/foca.py:
```diff
@@ -59,6 +64,12 @@ def foca(config: Optional[str] = None) -> App:
         )
     else:
         logger.info(f"No OpenAPI specifications provided.")
+        
```
Add comment describing what you are doing

kushagra189 replied:
done

uniqueg commented on foca/foca.py:
```diff
@@ -59,6 +64,12 @@ def foca(config: Optional[str] = None) -> App:
+    if conf.access and conf.access.enable:
+        cnx_app = _register_permission_specs(cnx_app)
+        logger.info(f"Permission management specs registered.")
+    else:
```
Probably enough to log once for the `else` case, something like `No access control configured.`

kushagra189 replied:
done

uniqueg commented on foca/foca.py:
```diff
@@ -70,6 +81,20 @@ def foca(config: Optional[str] = None) -> App:
     else:
         logger.info(f"No database support configured.")
 
+    # Register permission management and casbin enforcer
```
Couldn't the previous block from lines 68 to 72 maybe be moved down here so that we don't have three code blocks dealing with access control/permission management?

kushagra189 replied:
done

uniqueg commented on foca/permission_management/constants.py:
```diff
@@ -0,0 +1,4 @@
+"""File to store permission based constants
+"""
+PERMISSION_DB_NAME="access_db"
```
I know I said that these don't need to be configurable in the first implementation. Nevertheless I feel that they should be added to the config as everything will be in one place, there's less specific documentation to be written and one module less to maintain. If it's a hassle to make this configurable at the moment, you can just define them as attributes and not arguments and then hardcode these as defaults. However, if it's not a lot more work, you might as well make them configurable now.

In any case, we probably need to make a check somewhere that collection names within a database are not repeated. But that can go in a different PR (can create an issue)

kushagra189 replied:
Made them configurable.

uniqueg commented on foca/foca.py:
```diff
@@ -12,6 +12,8 @@
 from foca.factories.celery_app import create_celery_app
 from foca.security.cors import enable_cors
+from foca.permission_management.config_utils import _create_permission_config, _register_permission_specs, _register_casbin_enforcer
+from foca.permission_management.constants import PERMISSION_DB_NAME, PERMISSION_DB_COLLECTION_NAME
```
See the comment in the `constants` module about these

kushagra189 replied:
marking as done.

uniqueg commented on foca/foca.py:
```diff
@@ -70,6 +81,20 @@ def foca(config: Optional[str] = None) -> App:
+    if conf.access and conf.access.enable:
+        cnx_app = _register_casbin_enforcer(
+            app=cnx_app,
+            policy_path=conf.access.policy_path,
+            owner_headers=conf.access.owner_headers,
+            user_headers=conf.access.user_headers,
+            db_name=PERMISSION_DB_NAME,
+            collection_name=PERMISSION_DB_COLLECTION_NAME,
+            db_host=conf.db.host,
+            db_port=conf.db.port
```
I feel that this function has quite a messy API. At this time, all of these are available inside the `cnx_app` object (or, in the case of `PERMISSION_DB_NAME` and `PERMISSION_DB_COLLECTION_NAME` _should be_ in there, see other comments above and below in the `constants` module).

One thing about this though: I'm afraid in our current implementation of `register_mongodb()`, we have side effects, because we are basically just saying that we can pass a Connexion app object - but if people just do that without first attaching the config (including the right attributes/keys), they will get all kinds of `AttributeError`s.

So probably the best is to pass the minimum amount of config models (that's basically why we have them!) to the functions, but _not_ attributes individually. As a rule of thumb, I would apply the rule of threes here, i.e., if we need to pass three or more submodels to a function to pass all the attributes we are going to need in it, we should pass the next higher up submodel, all the way up to passing the entire `Config` model, if necessary. In this way, our functions' APIs and our main module here won't get too messy and we still have all the benefits of having explicit requirements with the possibility to check types. 

I would suggest to start implementing this pattern here (then later change it in `register_mongodb()` and other places, if there are more, please open an issue). So for registering the Casbin enforcer, the function signature would become something like this:

```
register_casbin_enforcer(
    app: App,
    ac_config: AccessControlConfig,
    db_config: MongoConfig,
) -> App:
```

kushagra189 replied:
done

uniqueg commented on foca/permission_management/constants.py:
```diff
@@ -0,0 +1,4 @@
+"""File to store permission based constants
```
Short docstrings <81 chars should go on one line as per PEP 8. Also, our docstrings end with a full stop. Anyway, we probably won't need this module, see comment below.

kushagra189 replied:
fixed as per Flake8

uniqueg commented on foca/permission_management/config_utils.py:
```diff
@@ -0,0 +1,124 @@
+"""Methods to manage permission management configuration"""
```
I would put this in `foca.access_control.register_casbin` to keep consistency with how we do it for `foca.database.register_mongodb`

kushagra189 replied:
done

uniqueg commented on requirements.txt:
```diff
@@ -6,11 +6,13 @@ Flask-Cors==3.0.9
 toml==0.10.0
 typing==3.7.4.1
 Werkzeug==1.0.1
+flask-authz==2.4.0
```
Please sort dependencies alphabetically

kushagra189 replied:
marking as done.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+openapi: 3.0.2
+info:
+  title: 'FOCA permission control API specification'
```
Better to use vocabulary consistently. Either `access control` or `permission management`. I'd favor the first one because it is shorter and is consistent with abbreviations like ABAC and RBAC.

kushagra189 replied:
marking as done.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+openapi: 3.0.2
+info:
+  title: 'FOCA permission control API specification'
+  description: 'CRUD and List Operations for managing permission.'
```
managing permissions (plural)

kushagra189 replied:
marking as done.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+  version: 1.0.0
+  license:
+    name: 'Apache 2.0'
+    url: 'https://raw.githubusercontent.com/ga4gh-discovery/ga4gh-service-info/develop/LICENSE'
```
Please link to the permissions in the FOCA repo instead

kushagra189 replied:
marking as done

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+    name: 'Apache 2.0'
+    url: 'https://raw.githubusercontent.com/ga4gh-discovery/ga4gh-service-info/develop/LICENSE'
+  contact:
+    name: 'GA4GH Discovery Networks Team'
```
Please use `ELIXIR Cloud & AAI` with the contact `alexander.kanitz@alumni.ethz.ch` (we really need a service email....)

kushagra189 replied:
marking as done.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+              schema:
+                $ref: '#/components/schemas/Error'
+servers:
+  - url: /admin/permission-management
```
better: `/admin/access-control`

kushagra189 replied:
marking as done.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+        id:
+          type: string
+          description: Unique id of the permission.
+        sec:
```
Better be explicit so that the specs will document themselves as best as possible.

kushagra189 replied:
Resolved.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+          type: string
+          description: Section name.
+          enum: ["p", "g"]
+        ptype:
```
Better: `policy_type`

kushagra189 replied:
resolved.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+          enum: ["p", "g"]
+        ptype:
+          type: string
+          description: Policy type.
```
What are the options here?

kushagra189 replied:
No options, anything can be defined.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+      properties:
+        id:
+          type: string
+          description: Unique id of the permission.
```
Better: "identifier" instead of "id"

kushagra189 replied:
resolved.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+          description: Unique id of the permission.
+        sec:
+          type: string
+          description: Section name.
```
Not obvious what `p` and `g` are. If possible, use explicit values here. Otherwise mention what the abbreviations stand for in the description.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+      description: Rule definition.
+      type: object
+      properties:
+        v0:
+          type: string
+          description: Param 0.
+        v1:
+          type: string
+          description: Param 1.
+        v2:
+          type: string
+          description: Param 2.
+        v3:
+          type: string
+          description: Param 3.
+        v4:
+          type: string
+          description: Param 4.
+        v5:
+          type: string
+          description: Param 5.
```
This is very opaque. Isn't there a better way to do this? If you want to allow users to provide an arbitrary set of keys and values, you can use `additionalProperties` like so:

```
Rule:
  description: Rule definition.
  type: object
  additionalProperties:
    type: string
```

See here for more details: https://swagger.io/docs/specification/data-models/dictionaries/

If you really really need these five properties, and each of them has a specific meaning, then it's best to give them descriptive names and extend the descriptions.

If you don't need to call the properties by name, then you should use a list of strings instead.

kushagra189 replied:
will explain alongside later.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+servers:
+  - url: /admin/permission-management
+components:
+  schemas:
```
All the schemas should include reasonable examples: https://swagger.io/docs/specification/adding-examples/

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Permission'
+        '404':
```
Should contain `401` and `500` as well, and `404` should be listed _after_ `403`.

kushagra189 replied:
marking as done.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/Permission'
```
Add errors responses: `400`, `401`, `403` and `500`

kushagra189 replied:
marking as done.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+      description: Returns all permissions existing on the system.
+      tags:
+        - permission
+      parameters:
```
This will not be enough to allow pagination, but I think for now that's fine (but please open an issue).

kushagra189 replied:
done.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+              schema:
+                description: Identifier of deleted permission.
+                type: string
+        '400':
```
`400` responses in `GET` and `DELETE` endpoints where only path parameters are defined and adding arbitrary query parameters is not forbidden probably cannot arise (I'm not sure). I'm fine with removing them in these cases, but I'm also fine with keeping them in the specs just in case (they won't hurt). However, let's try to be consistent, so please either add them for both the `GET` and `DELETE` endpoints, or remove them for both.

kushagra189 replied:
marking as done.

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+    get:
+      summary: List permissions
+      operationId: foca.permission_management.permission_server.getAllPermissions
+      description: Returns all permissions existing on the system.
```
This reminds me of a potential feature (perhaps you can open an issue): there could be a config option that allows users to check their own permissions. In that case the `GET` endpoints should be available to the users for any permissions that affect them.

kushagra189 replied:
done.

uniqueg commented on foca/permission_management/permission_server.py:
```diff
@@ -0,0 +1,160 @@
+
+import logging
+
+from typing import Dict, List
```
Please use tuple notation for multiple imports from the same module

kushagra189 replied:
Done

uniqueg commented on foca/permission_management/permission_server.py:
```diff
@@ -0,0 +1,160 @@
+import logging
+
+from typing import Dict, List
+from bson.objectid import ObjectId
```
I think `bson` is not a built-in module so should be grouped with `flask` in the next block

kushagra189 replied:
done

uniqueg commented on foca/permission_management/permission_server.py:
```diff
@@ -0,0 +1,160 @@
+
+from foca.utils.logging import log_traffic
+from foca.permission_management.constants import PERMISSION_DB_COLLECTION_NAME, PERMISSION_DB_NAME
+from werkzeug.exceptions import InternalServerError, NotFound
```
This should be grouped with `flask` and `bson` in the previous block.

Also, please sort each import bock alphabetically, by module name (so `foca.permission_management` _before_ `foca.utils`, although the `PERMISSION_DB_` params should end up in `current_app` anyway)

kushagra189 replied:
Done

uniqueg commented on foca/permission_management/permission_server.py:
```diff
@@ -0,0 +1,160 @@
+        rule = request_json.get("rule", {})
+        permission_data = {
+            "ptype": request_json.get("ptype", None) if request_json.get("ptype", None) else None,
+            "v0": rule.get("v0", None) if rule.get("v0", None) else None,
```
This is really really cryptic and I have no clue what's happening with these permissions. Do you have a reference where the use of `v0`, `v1` etc is explained? This is really quite problematic if we have to keep those names. I'm sure there will be a more explicit/descriptive name

uniqueg replied:
Also, I don't understand the use of the ternary operator here: we set key `v0` to the value of `v0` in `rule`, if it exists, otherwise to `None`, right? But this is already achieved with this:

```python
"v0": rule.get("v0", None)
```

No need for anything that follows, I think

kushagra189 replied:
resolved

uniqueg commented on foca/permission_management/permission_server.py:
```diff
@@ -0,0 +1,160 @@
+            "v3": rule.get("v3", None) if rule.get("v3", None) else None,
+            "v4": rule.get("v4", None) if rule.get("v4", None) else None,
+            "v5": rule.get("v5", None) if rule.get("v5", None) else None
+        }
```
So where is the `sec` property used?

kushagra189 replied:
Fixed

uniqueg commented on foca/permission_management/api/permission-endpoint-specs.yaml:
```diff
@@ -0,0 +1,277 @@
+          type: string
+          description: Policy type.
+        rule:
+          $ref: '#/components/schemas/Rule'
```
Might perhaps be useful to include an optional `name` property? Possibly also for the rule?

kushagra189 replied:
Not required.

uniqueg commented on foca/permission_management/permission_server.py:
```diff
@@ -0,0 +1,160 @@
+        }
+        permission = db_coll_permission.insert_one(document=permission_data)
+        logger.info("Policy added.")
+        return str(permission.inserted_id)
```
Aren't the identifiers that MongoDB returns really unwieldy? Would perhaps be better to generate more managable identifiers. Makes that controller a lot longer, but the code be copied largely from TRS-Filer etc.

kushagra189 replied:
need to check

uniqueg commented on foca/permission_management/permission_server.py:
```diff
@@ -0,0 +1,160 @@
+""""Controllers for permission management endpoints."""
```
What's missing entirely here is the handling of `401` (should be handled automatically if you add the security definitions via `SpecConfig) and the handling of `403`.

Also, we should have a default (possibly immutable) policy in place for at least one `admin` user if `auth` is enabled.

kushagra189 replied:
resolved.

uniqueg commented on foca/permission_management/config_utils.py:
```diff
@@ -0,0 +1,124 @@
+    
+    app.app.config['casbin_enforcer'] = casbin_enforcer
+    app.app.config['casbin_adapter'] = adapter
+    return app
```
Add newline

kushagra189 replied:
done

uniqueg commented on foca/permission_management/permission_server.py:
```diff
@@ -0,0 +1,160 @@
+    if del_obj_permission.deleted_count:
+        return id
+    else:
+        raise NotFound
```
Add newline

kushagra189 replied:
marking as done.

uniqueg commented on foca/permission_management/config_utils.py:
```diff
@@ -0,0 +1,124 @@
+    return app
+
+
+def _register_casbin_enforcer(
```
See other comment ablove re: making this API less crowded

kushagra189 replied:
Done.

uniqueg commented on foca/permission_management/config_utils.py:
```diff
@@ -0,0 +1,124 @@
+        path=str(resource_filename("foca.permission_management", "api/permission-endpoint-specs.yaml")),
+        add_operation_fields={
+            "x-openapi-router-controller": "foca.permission_management.permission_server"
+        },
```
We need to add security definitions as well!

uniqueg commented on foca/permission_management/config_utils.py:
```diff
@@ -0,0 +1,124 @@
+    return config
+
+
+def _register_permission_specs(
```
Don't we have a mechanism for merging YAML config files in here somewhere? If so, we could just put the config for these paths in a YAML file and merge it with/append it to the app config that the user provides. Then we could get rid of this function entirely and the user could (if they really know what they are doing) easily modify that, e.g., `config_access_control.yaml` file.

kushagra189 replied:
will discuss the current implementation.

uniqueg commented on foca/permission_management/config_utils.py:
```diff
@@ -0,0 +1,124 @@
+                        )
+                    }
+                else:
+                    # TODO: add check so user cannot enter access_db as the db name.
```
Yes, please implement that or create an issue and then remove this `TODO`

uniqueg replied:
Or actually, is there a reason why they couldn't store it in the same database (but a different collection)? So perhaps we should just give a warning if they choose to add an `access_control_db`, but only error if both the db _and_ collection name are not unique

kushagra189 replied:
Not a problem we can use this. Creating an issue for now.

uniqueg commented on foca/permission_management/config_utils.py:
```diff
@@ -0,0 +1,124 @@
+    if config.access is not None:
+        if config.access.enable:
+            # Add index for permission management
+            if config.db is not None:
```
And what if it _is_ `None`? Even if the user doesn't set up a db for the app - if they want to use our access control, we will need to tell the user to include config params for `MongoConfig`

kushagra189 replied:
not needed anymore

uniqueg commented on foca/permission_management/config_utils.py:
```diff
@@ -0,0 +1,124 @@
+                    }
+                else:
+                    # TODO: add check so user cannot enter access_db as the db name.
+                    config.db.dbs['access_db'] = DBConfig(collections=None, client=None)
```
`'access_db'` shouldn't be hard-coded here, I think? And why don't we assign a collection name here?

kushagra189 replied:
removed.

uniqueg commented on foca/permission_management/config_utils.py:
```diff
@@ -0,0 +1,124 @@
+            # Add index for permission management
+            if config.db is not None:
+                if config.db.dbs is not None:
+                    config.db.dbs = {
```
Do we really wanna override the databases if there are no databases (`not None`)? Shouldn't it rather be the other way around: if there are already databases, we add another one, and if there aren't we just define it?

kushagra189 replied:
fixed implementation.

uniqueg commented on foca/permission_management/config_utils.py:
```diff
@@ -0,0 +1,124 @@
+    Returns:
+        Updated application configuration.
+    """
+    if config.access is not None:
```
This code is quite messy and complicated. I'm sure that can be solved more elegantly.

kushagra189 replied:
fixed

uniqueg commented on foca/permission_management/config_utils.py:
```diff
@@ -0,0 +1,124 @@
+logger = logging.getLogger(__name__)
+
+
+def _create_permission_config(
```
Probably we should do the same approach I suggested below: We provide a `database:` section in the `config_access_control.yaml` and then merge/append that file to the user's app config - in this way we would for sure have at least one database defined and we could just use the already existing mechanism to add this database and collection as well...

kushagra189 replied:
done

uniqueg reviewed (changes requested):
After all, there are quite a lot of comments. Most of them are very simple to fix, but some need some re-design.

The most critical issues:
- The permission/rule model in the spec still needs some work - see the individual comments. Currently I am not able to follow how permissions are really constructed based on these `v0` through `v5` properties. At the very least this will need much better documentation, but possibly also some redesign. It's probably best to start with this. Please describe in some detail how this currently works, why you have implemented it with this `v0` to `v5` pattern and then we can see how we can improve on that.
- The new controllers need handling of `401` (should be handled automatically if `auth` is `required` and security definitions are added in the access control `SpecConfig`) and `403` (implement in the controllers themselves), and the default policy file has to attach all permissions to a single admin user who can then manage permissions via the API (this needs to be configured and implemented somehow still, I think).
- I think we should use the already existing mechanisms in FOCA to set up the access control database and API endpoint registration via a config file. This will make the main module a lot easier and leave just the Casbin enforcer registration in there. I think what we could, for example, have the following behavior: we remove the `enable` property and configure access control whenever the user provides at least the `access_control:` stub. Inside the `AccessControlConfig` model we could provide two additional properties `use_default_db_config` and `use_default_api_specs`. For each of these, when set to `True` (default), the config files are read from a file `config_defaults_access_control.yaml` inside the FOCA repo where the necessary default `db.dbs` and `api.spec` config params are listed, and appended to the app config provided by the user. If set to `False`, users are responsible for providing suitable `db.dbs` and `api.specs` entries in the app config themselves. The `AccessControlConfig` should also include the `db_name` and `collection_name` properties, which, by default, should match the default access control database configuration in `config_defaults_access_control.yaml`. Other properties in the `AccessControlConfig` can remain as they are. In this way, the minimum that's needed from a user to enable access control with the default configuration is just the `access_control:` stub without any params passed to any of the properties (as they all have defaults). If they _don't_ want access control, they can just omit the `access_control:` section altogether. In this way, we only need a single function to register Casbin, the policy file, and a file containing the default specs. The config handler has to be changed to accommodate the merging behavior if the default access control config is to be used, including the check for database/collection name consistency between what is provided as db and collection names in the `AccessControlConfig` and the databases/collections that are to be registered with MongoDO _and_ ensuring that no two collections of the same name are to be registered in the same database (it should be fine to have the same collection name in different databases and different collections in the same database though). This latter check should be applied across all listed database/collection pairs, not just between the ones for access control and the ones provided by the user. Also, this will probably need to be done prior to merging to ensure that a user entry is not overridden by merging or vice versa. Again, all these changes need to be done in the config parser, in the most generic way possible.
- Make sure that the petstore test really works with the behavior described in the points above, including the admin user. This will likely require not only `access_control` to be enabled but also `auth.require` to be set to `True` as we will need to have a user identity passed to the endpoints, or not? But it's a good idea anyway to have a petstore example with authorization configured for all endpoints. However, I would recommend to have a second petstore example `petstore_access_control` for this case, leaving the current one as it is (or was before this PR).
- Finally, I think it's time to set the default for `auth.require` to `True` in https://github.com/elixir-cloud-aai/foca/blob/fb2cd6573f346125be6b1c312a6c344c87b87614/foca/models/config.py#L705 - we don't want to encourage people to develop unprotected APIs! Speaking of this, it seems that we forgot to document the `require` property in that model, in the `Args:`, the `Attributes:` and the `Example:` sections: https://github.com/elixir-cloud-aai/foca/blob/fb2cd6573f346125be6b1c312a6c344c87b87614/foca/models/config.py#L650-L698, also see here: https://foca.readthedocs.io/en/latest/modules/foca.models.html#foca.models.config.AuthConfig. Let's use the opportunity and add that.

I think it shouldn't be as much as it sounds. I suppose the biggest work was to understand casbin and get to a first implementation, and for that it's pretty good. Quite a bit of the suggestions are basically just refactoring. But given that we want to have a good version before we move to Anurag and that this affects the behavior of FOCA deeply, and currently in (too) many modules, I think it's quite important that we implement this as clean as possible in the first round. I don't think this PR is too far awat from that :)

kushagra189 committed:
fix review comments

kushagra189 committed:
fix req

kushagra189 committed:
fix req

kushagra189 committed:
fix req

uniqueg committed:
Update docker-compose.yaml

uniqueg commented on examples/petstore/config.yaml:
```diff
@@ -42,3 +42,7 @@ db:
+
+access_control:
+    use_default_db_config: True
+    use_default_api_specs: True
```
See comment further down when model is defined.

uniqueg commented on foca/models/config.py:
```diff
@@ -1094,6 +1096,93 @@ class LogConfig(FOCABaseConfig):
+        policy_path: Path to access control model configuration file.
+        owner_headers: Owner (Admin) specific header property requirements
+            for casbin.
+        user_headers: User specific header property requirements for casbin.
```
Not a big fan of this design, for the following reasons:
- If not using the defaults DB config and API specs, multiple options need to be specified. This makes the class API unnecessarily complex. Would be better to, e.g., set `api_specs_path` to `None` if the default specs are to be used. Also, the behavior for setting the flags to `True` but indicating a specs path and/or a DB config is ill-defined and could cause confusion (which ones are taken?). This would not be an issue when using a single option that can be set to `None` when defaults are desired. Document this though in the arg description, but not the attribute description (the attribute should never be `None` and always be a path).
- It is unclear what the "DB config" is, just db and collection name? Also, again, using defaults is tricky here: what if the user says "True" but provides a value for `db_name` but not for `collection_name`? And what if that's actually what they want - use the default collection name but specify the db name explicitly? Again, removing the flag and allowing `db_name` and `collection_name` to be set to `None` to use defaults is more robust. However, again, document this behavior in the arg (but not attribute) description.
- Use consistent naming: `api_specs_path` vs `api_spec_controller_path` - use either `spec` or `specs`. Also, `path` does not need to be included, IMO, especially if it can take on the value of `None`. Just `api_specs` and `api_controllers` would do, IMO. And if you go with plural here, I would also change `policy_path` to `policies` (again, no `path`). See next comment for more considerations on paths.
- Consistent with Connexion, it might make sense to use `package.module` notation, rather than `/path/module` notation.
- The documentation for `owner_headers` and `user_headers` is not detailed enough - judging by the fact that I don't understand it. We are hiding complexity here, so mentioning casbin is not very helpful. Write documentation such that a user will know what to put here.

uniqueg reviewed (changes requested):
@kushagra189: the PR is too big and reviewing it I'm getting caught up in minor issues. I think we can probably leave it as is (maybe address the few comments I have made) and then refactor things later.

However, what I would really like to see before merging this is detailed documentation of how FOCA deals with access control and the petstore example. So we need a section in `README.md` that should include *detailed* instructions/explanations on:

- the FOCA access control configuration (default and how they can be customized)
- specs and controllers for access control management (default and how they can be customized)
- the policy file (petstore example and how to customize it)
- how to run the petstore app with access control enabled
- how to test that access control works as intended given the default policies (i.e., instructions on how to actually see it in action)
- how to manage access conrol given the default specs and controllers (including actual example API calls)

After reading this, a user should be able to see access control in action by deploying the petstore app localy, understand the mechanism by which FOCA enables configuring and managing access control, understand the default behavior and have enough insight to modify the defaults according to desired behavior.

kushagra189 committed:
fix review comments

kushagra189 committed:
fix review comments

kushagra189 committed:
Merge branch 'feature/add_access_control' of https://github.com/elixir-cloud-aai/foca into feature/add_access_control

kushagra189 committed:
added example schemas

kushagra189 committed:
added security def

kushagra189 committed:
fix req merge

kushagra189 committed:
fix example schema

kushagra189 committed:
fix merge

kushagra189 committed:
fix merge

kushagra189 committed:
added docs

kushagra189 committed:
fix merge

kushagra189 committed:
fix checks

kushagra189 committed:
added decorator for check permissions

kushagra189 committed:
update readme

kushagra189 committed:
fix flake8 issues

kushagra189 committed:
fix existing test cases

kushagra189 committed:
added endpoint definition in readme

kushagra189 committed:
fix foca tests

kushagra189 committed:
add tests for casbin rule objects

kushagra189 committed:
fix flake8

kushagra189 committed:
added tests for adapter

kushagra189 committed:
fix flake8

kushagra189 committed:
add mongodb startup to github actions

kushagra189 committed:
add mongodb startup to github actions

kushagra189 committed:
add mongodb stop to github actions

kushagra189 committed:
add support on custom port

kushagra189 committed:
added tests for endpoints

kushagra189 committed:
fix db config for tests

kushagra189 committed:
added tests for post and put controller

kushagra189 committed:
fix db dependency

kushagra189 committed:
update files

kushagra189 committed:
cleaned test case files

kushagra189 committed:
fix flake 8

kushagra189 committed:
fix tests

kushagra189 committed:
fix port

kushagra189 committed:
complete test coverage

uniqueg committed:
feat: env var to select Py image for Petstore

Change documented in the Petstore README.md.

Also fixed a permission issue arising during Docker build, because the
Petstore docker-compose configuration creates the database directory
inside the Docker build context, which is not accessible as it is
created by user `root`. This affects the builds of both the FOCA and the
Petstore image and causes the Petstore app build to fail if the database
directory already exists (e.g., when rebuilding after the first build).
The `examples/` and `data/` directories were now added to
`.dockerignore` files for FOCA and Petstore builds respectively.

Finally, removed some extra whitespace in two files.

kushagra189 committed:
fix example config for petstore

kushagra189 committed:
added test application for access control

kushagra189 committed:
fix config definition

kushagra189 committed:
fix access control app

kushagra189 committed:
fix test cases

kushagra189 committed:
fix controllers

kushagra189 committed:
fix yaml file config

uniqueg committed:
tests(petstore-auth): fix 401 and 403 responses (#136)

kushagra189 committed:
fix tests

kushagra189 committed:
fix integration tests

kushagra189 committed:
fix integration tests

kushagra189 committed:
fix controller actions for permissions

uniqueg committed:
docs: update main & access control docs (#138)

uniqueg commented:
🚀
